src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2874,7 +3149,12 @@ public <T> ForkJoinTask<T> submit(ForkJoinTask<T> task) {
      */
     @Override
     public <T> ForkJoinTask<T> submit(Callable<T> task) {
-        return poolSubmit(true, new ForkJoinTask.AdaptedCallable<T>(task));
+        ForkJoinTask<T> t =
+            (Thread.currentThread() instanceof ForkJoinWorkerThread) ?
```

Comment:
@DougLea Won't we need to check if the submitting thread belongs to the same pool here?

Reply:
No, but I can imagine this being the subject of a Puzzler someday, because either choice might be unexpected by someone! To maximize compatibility wrt incompatible specs (initial FJ and ES), calls to submit from workers are by default noninterruptible, even if they are client calls into another pool. The opposite choice might be more defensible but could break existing code.

Reply:
@DougLea Ok, that's good to know. :+1: